dbms/src/Common/ConcurrentIOQueue.h

```diff
@@ -0,0 +1,219 @@
+            case MPMCQueueResult::OK:
+                pushRemainsWithoutLock();
+            default:
+                return res;
+            }
+        }
+        case MPMCQueueResult::OK:
+            pushRemains();
+        default:
+            return res;
+        }
+    }
+
+    size_t size() const
+    {
+        size_t mpmc_queue_size = mpmc_queue.size();
+        std::lock_guard lock(mu);
+        return mpmc_queue_size + remaining_objs.size();
+    }
+
+    // When the queue is finished, it may appear that the mpmc_queue is empty but the remaining_objs are not, causing isFull to return true.
+    // But this is expected, and the return value of isFull is meaningless after finished/cancelled.
+    bool isFull() const
+    {
+        {
+            std::lock_guard lock(mu);
+            if (!remaining_objs.empty())
+                return true;
+        }
+        return mpmc_queue.isFull();
+    }
+
+    MPMCQueueStatus getStatus() const
+    {
+        return mpmc_queue.getStatus();
+    }
+
+    /// Cancel a NORMAL queue will wake up all blocking readers and writers.
+    /// After `cancel()` the queue can't be pushed or popped any more.
+    /// That means some objects may leave at the queue without poped.
+    bool cancel()
+    {
+        return cancelWith("");
+    }
+    bool cancelWith(String reason)
+    {
+        if (mpmc_queue.cancelWith(std::move(reason)))
+        {
+            std::lock_guard lock(mu);
+            remaining_objs.clear();
+            return true;
+        }
+        return false;
+    }
+
+    const String & getCancelReason() const
+    {
+        return mpmc_queue.getCancelReason();
+    }
+
+    /// Finish a NORMAL queue will wake up all blocking readers and writers.
+    /// After `finish()` the queue can't be pushed any more while `pop` is allowed
+    /// the queue is empty.
+    /// Return true if the previous status is NORMAL.
+    bool finish()
+    {
+        return mpmc_queue.finish();
+    }
+
+private:
+    void pushRemainsWithoutLock()
+    {
+        while (!remaining_objs.empty() && mpmc_queue.tryPush(std::move(remaining_objs.back())) == MPMCQueueResult::OK)
+            remaining_objs.pop_back();
+    }
+
+    void pushRemains()
+    {
+        std::lock_guard lock(mu);
+        pushRemainsWithoutLock();
+    }
+
+    bool popRemainsWithoutLock(T & data)
+    {
+        if (!remaining_objs.empty())
+        {
+            data = std::move(remaining_objs.back());
+            remaining_objs.pop_back();
+            return true;
+        }
+        return false;
+    }
+
+    bool popRemains(T & data)
+    {
+        std::lock_guard lock(mu);
+        return popRemainsWithoutLock(data);
+    }
+
+private:
+    mutable std::mutex mu;
+    // Used to hold objs pushed by `NonBlockingPush` that exceed the capacity.
+    // pop/tryPop will try to push remaining_objs into the mpmc_queue when pop/tryPop.
+    std::deque<T> remaining_objs;
```

Comment:
Will this `deque` actually makes `ConcurrentIOQueue` unbounded?

Reply:
No, because the way of using is
```
while (queue.isFull()) {}
queue.nonBlockingWrite(block);
```
The size of the deque is less than or equal to the number of write threads.

Reply:
If you use the blocking write method, the thread will also hold the obj on the thread stack.

Reply:
If you use the nonBlockingWrite as 
```
while (queue.isFull()) {}
queue.nonBlockingWrite(block);
```
this piece of code itself is blocking?

Reply:
In the pipeline model, queue.isFull() is executed in the `WaitReactor`, so it does not block the compute thread.

Reply:
like
```
struct AsyncWriter : public SinkOp
{
void await()
{
    return queue.isFull() ? Status::Waiting : Status::NEED_INPUT;
}
void write(Block && block)
{
    queue.nonBlockingWrite(std::move(block));
}
}
```

Reply:
OK